lib/puppet/util/character_encoding.rb

```diff
@@ -0,0 +1,96 @@
+          # fail with Encoding::InvalidByteSequenceError, referencing the
+          # invalid byte sequence in the original, pre-transcode, string. We
+          # might have gotten here, for example, if puppet is run first in a
+          # user context with UTF-8 encoding (setting the "is" value to UTF-8)
+          # and then later run via cron without UTF-8 specified, resulting in in
+          # EN_US (ISO-8859-1) on many systems. In this scenario we're
+          # effectively best-guessing this string originated as UTF-8 and only
+          # set external encoding to UTF-8 - transcoding would have failed
+          # anyway.
+
+          # The second scenario (more rare, I expect) is that this string does
+          # NOT have invalid byte sequences (string.valid_encoding? == true),
+          # but is *ALSO valid unicode*.
+          # Our example case is "\u16A0" - "RUNIC LETTER FEHU FEOH FE"
+          # http://www.fileformat.info/info/unicode/char/16A0/index.htm
+          # 0xE1 0x9A 0xA0 / 225 154 160
+          # These bytes are valid in ISO-8859-1 but the character they represent
+          # transcodes cleanly in ruby to *different* characters in UTF-8.
+          # That's not what we want if the user intended the original string as
+          # UTF-8. We can only guess, so if the string is valid UTF-8 as
+          # currently constituted, we default to assuming the string originated
+          # in UTF-8 and do not transcode it - we only set external encoding.
+          value_to_encode.force_encoding(Encoding::UTF_8)
+        elsif transcodable?(value_to_encode)
+          # If the string is not currently valid UTF-8 but it can be transcoded,
+          # we can guess this string was not originally unicode. We need it to
+          # be now, so transcode it to UTF-8. For strings with original
+          # encodings like SHIFT_JIS, this should be the final result.
+          value_to_encode.encode!(Encoding::UTF_8)
+        else
+          # If the string is neither valid UTF-8 as-is nor is transcodable, fail
+          # at this point. It requires user remediation.
+          raise EncodingError
+        end
+      rescue EncodingError => detail
+        # catch both our own self-determined failure to transcode as well as any
+        # error on ruby's part, ie Encoding::InvalidByteSequenceError or
+        # Encoding::UndefinedConversionError.
+        raise Puppet::Error, _("#{detail.inspect}: #{value_to_encode.dump} is not valid UTF-8 and cannot be transcoded by Puppet.")
+      end
+
+      return value_to_encode
+    end
+
+    private
+
+    # Do our best to determine if a string is valid
+    # UTF-8 via String#valid_encoding?
+    # @api private
+    # @param [String] string a string to test
+    # @return [Boolean] whether we think the string is UTF-8 or not
+    def valid_utf_8?(string)
+      string.dup.force_encoding(Encoding::UTF_8).valid_encoding?
```

Comment:
I'm not sure about this implementation, as if it becomes pervasive, this will generate quite a bit of string data on the Ruby heap that isn't really used, which could lead to perf issues.

If we could do an in-place scan against the bytes, that would be better.

* I checked for another implementation of `valid_encoding?` elsewhere in the Ruby code, and didn't see anything very obvious - under the hood `rb_enc_str_coderange` is called.  That function lives at https://github.com/ruby/ruby/blob/7de42daa21a47335a485265f74322533a2da4f1a/string.c#L619-L637
* One hacky way of just using Ruby would be to save the encoding, set encoding to UTF-8, check `valid_encoding?` and then reset back to the original.  Gross.
* UTF-8 bytes sequences are actually pretty easy to detect by scanning strings byte-by-byte, so we could write a quick implementation that does just that in-place on the existing string.  I was going to point you to the Wikipedia page, but I found a quick example of this at http://codereview.stackexchange.com/a/59439


Also, some more related discussion on detecting UTF-8 for reference:
https://stackoverflow.com/questions/8635578/how-to-check-whether-the-character-is-utf-8